Mathlib/Geometry/Manifold/LocalDiffeomorph.lean

```diff
@@ -0,0 +1,159 @@
+because structomorphisms do not allow the model spaces `H` and `H'` of the two manifolds to be
+different, i.e. for a structomorphism one has to impose `H = H'` which is often not the case in
+practice.
+
+## Tags
+local diffeomorphism, manifold
+
+-/
+
+open Manifold Set  TopologicalSpace
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+  {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+  {H : Type*} [TopologicalSpace H]
+  {G : Type*} [TopologicalSpace G]
+  (I : ModelWithCorners 𝕜 E H) (J : ModelWithCorners 𝕜 F G)
+  (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+  (N : Type*) [TopologicalSpace N] [ChartedSpace G N] (n : ℕ∞)
+
+section LocalDiffeomorphAux
+/-- A "diffeomorphism on" `s` is a function `f : M → N` such that `f` restricts to a diffeomorphism
+`s → t` between open subsets of `M` and `N`, respectively.
+This is an auxiliary definition and should not be used outside of this file. -/
+structure LocalDiffeomorphAux extends LocalHomeomorph M N where
+  contMDiffOn_toFun : ContMDiffOn I J n toFun source
+  contMDiffOn_invFun : ContMDiffOn J I n invFun target
+
+/-- Coercion of a `LocalDiffeomorphAux` to function.
+Note that a `LocalDiffeomorphAux` is not `FunLike` (like `LocalHomeomorph`),
+as `toFun` doesn't determine `invFun` outside of `target`. -/
+instance : CoeFun (LocalDiffeomorphAux I J M N n) fun _ => M → N :=
+  ⟨fun Φ => Φ.toFun'⟩
+
+/-- A diffeomorphism is a local diffeomorphism. -/
+def Diffeomorph.toLocalDiffeomorphAux (h : Diffeomorph I J M N n) : LocalDiffeomorphAux I J M N n :=
+  {
+    toLocalHomeomorph := h.toHomeomorph.toLocalHomeomorph
+    contMDiffOn_toFun := fun x _ ↦ h.contMDiff_toFun x
+    contMDiffOn_invFun := fun _ _ ↦ h.symm.contMDiffWithinAt
+  }
+
+-- Add the very basic API we need.
+namespace LocalDiffeomorphAux
+variable (Φ : LocalDiffeomorphAux I J M N n) (hn : 1 ≤ n)
+protected theorem contMDiffOn : ContMDiffOn I J n Φ Φ.source :=
+  Φ.contMDiffOn_toFun
+
+protected theorem contMDiffOn_symm : ContMDiffOn J I n Φ.invFun Φ.target :=
+  Φ.contMDiffOn_invFun
+
+protected theorem mdifferentiableOn : MDifferentiableOn I J Φ Φ.source :=
+  (Φ.contMDiffOn).mdifferentiableOn hn
+
+protected theorem mdifferentiableOn_symm : MDifferentiableOn J I Φ.invFun Φ.target :=
+  (Φ.contMDiffOn_symm).mdifferentiableOn hn
+
+protected theorem mdifferentiableAt {x : M} (hx : x ∈ Φ.source) : MDifferentiableAt I J Φ x :=
+  (Φ.mdifferentiableOn hn x hx).mdifferentiableAt (Φ.open_source.mem_nhds hx)
+
+-- define symm just to make this easier to write?
+protected theorem mdifferentiableAt_symm {x : M} (hx : x ∈ Φ.source) :
+    MDifferentiableAt J I Φ.invFun (Φ x) :=
+  (Φ.mdifferentiableOn_symm hn (Φ x) (Φ.map_source hx)).mdifferentiableAt
+  (Φ.open_target.mem_nhds (Φ.map_source hx))
+
+/- We could add lots of additional API (following `Diffeomorph` and `LocalHomeomorph*), such as
```

Comment:
It would be nice to at least define `symm`, and state `Φ.invFun` as `Φ.symm` (with a coercion) everywhere. That's how we do it for all other equivalence-like structures.